rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+# 004 - Better Policy UX RFC
+
+## Current Status
+
+### Proposed
+
+2022-09-06
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Joseph <joseph@iasql.com>
+
+## Summary
+
+Customers (even the very technically experienced ones) often have to look up how to construct an AWS IAM policy and then spend time validating the policy before inserting into IaSQL.
+This becomes more and more unmanageable as the policy grows.
+This is not a good experience, and it can be improved upon. We believe we can collect some inputs from IaSQL customers and offer then some form of automation when creating policies, so they can do it faster, more confidently and more conveniently.
+
+## Proposal
+
+We should add new `Policy` and `Statement` entities (perhaps in the `aws_iam` module). These entities would serve as 
+a way to construct IAM policies in a more user-friendly way. For example, the customer intends to create an S3 bucket with
+read-only permissions from all Principals within the same AWS account. Ideally, something like this:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Resource": "aws:arn:123456789012:bucket-resource",
+      "Action": [
+        "s3:GetObjects"
+      ],
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:root"
+      }
+    }
+  ]
+}
+```
+
+Instead of the customer having to look up and put together a JSON blob like this and then verifying and then inserting, we can also give them
+the option to do this using functions that would generate this blob.
+
+```sql
+-- User creates policy
+select * from iasql_create_policy(
+    'some_policy_name',
+    '2012-10-17' -- nullable, defaults to 2012-10-17
```

Comment:
If the policy document version is changed to another version, how will the tables handle the changes to the JSON properties allowed?

Reply:
There are only 2 versions right now (2008 and 2012) and AWS strongly warns against using the 2008 one for new/existing policies.
https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_version.html

AFAIK, both versions support the same JSON properties, but the old version is missing some features like policy variables.
Case: If the user sets a version as `2008-10-17` and sets the `Principal` as `${aws:username}`, the policy is generated as-is still. Though, on `apply`, they will get the regular AWS error saying "Principal does not exist", because version `2008-10-17` will read the variable as a string literal.

What specific case do you envision?